torch/nested/_internal/cached_tensor.py

```diff
@@ -0,0 +1,131 @@
+from contextlib import contextmanager
+from typing import *  # noqa: F403
+
+import torch
+from torch.utils import _pytree as pytree
+
+
+class CachedTensor(torch.Tensor):
+    metadata: Dict[str, Optional[torch.Tensor]]
+    source_field: str
+
+    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
+    # is determined by a "source" tensor in the dict (specified by the user
+    # during construction).
+    #
+    # This class is not super useful on its own because by default, performing any
+    # operations on it will be as if you first unwrapped the CachedTensor and then
+    # performed the operation on the plain source tensor (a plain tensor is also returned).
+    # To leverage the extra metadata, you must register an op to perform the special logic
+    # you want via register_cached_tensor_func.
+    #
+    # When used this way (1) it is a convenient way to keep around metadata
+    # related to a tensor, without having to laboriously thread those extra metadata
+    # around, e.g. through aten signatures. (2) allows one to trigger custom
+    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
+    # otherwise because the subclass constructor op itself usually does not take the
+    # subclass itself as input!). See NestedTensor for an example.
+
+    @staticmethod
+    @torch._disable_dynamo  # type: ignore[misc]
+    def __new__(
+        cls,
+        metadata: Dict[str, Optional[torch.Tensor]],
+        source_field: str,
+    ) -> "CachedTensor":
```

Comment:
Okay I'm loving how much simpler this class is now! There's still one thing bothering me (from the perspective of reviewing this in isolation without looking much in detail at how it's used for NJT): the concept of a specifying a dict and source field that refers to an object in that dict seems overly restrictive and a bit awkward.

Ideally I'd want to see something like:
```
CacheTensor(t: torch.Tensor, cache: Dict[str, Optional[torch.Tensor]])
```

where the `CacheTensor` operates like `t` but also contains the extra `cache` stuff alongside it, and notably `cache` doesn't contain `t`. This seems about as clean as possible.

Would this cause problems with the way NJT wants to think about `source_field`? It feels like this should still accomplish what NJT needs, albeit slightly differently, but lmk if you think there are unsolvable perf issues with this.

Reply:
The main difference I see between the two approaches are that one requires the source field to have a name, the other does not. 

For NJT requiring the source field to have a name is useful because it allows me to access every field on the CacheTensor including the source field in a uniform way.

Reply:
Hm I'm not super convinced, especially because I want to consider this class outside of an NJT context. The source field thing makes CacheTensor usage marginally simpler within NJT but more awkward for other use cases imo. Treating the source field and some other cached objects uniformly seems pretty specific to the NJT use case, and arguably should be handled at the NJT layer.

> For NJT requiring the source field to have a name is useful because it allows me to access every field on the CacheTensor including the source field in a uniform way.

You can maintain a name within NJT, no? e.g. store canonical source field for a given NJT instance:
```
class NJT:
    ...
    source_field = "_device_offsets"  # for example
    source = CacheTensor(source_tensor, other_tensors)

    # custom getattr returns CacheTensor's t for source_field and indexes other_tensors for other fields.
    # it's less uniform from the perspective of NJT usage but places responsibilities at the correct layer imo
```

From what I can tell, this isn't too hard to deal with in NJT but it does make CacheTensor even dumber and straightforward (in a good way imo)

Reply:
Is the concern is that CacheTensor in its current state isn't constructed like one would expect a tensor named CacheTensor? e.g. in this model one would expect CacheTensor to be first and foremost a tensor, but has additional fields stored in a dict. Given this, I think the concern about its construction makes sense.

Could I alternatively propose that we changed the name of the tensor to `SimpleTensorDict`? In that model this tensor is more like a wrapper over a dict, in which every field is seen uniformly. In this world, the current way of construction should be most natural.

Reply:
Thinking out side of the NJT context, perhaps a reason to favor `SimpleTensorDict` over `CacheTensor` is that
CacheTensor doesn't actually benefit much from being first and foremost a tensor, because it immediately unwraps upon any operators call.  A container-like name of  simple-tensor-dict makes it clearer that this is more of a container object than a tensor you are meant to operate on.

Reply:
> Is the concern is that CacheTensor in its current state isn't constructed like one would expect a tensor named CacheTensor? e.g. in this model one would expect CacheTensor to be first and foremost a tensor, but has additional fields stored in a dict. Given this, I think the concern about its construction makes sense.

Yep you nailed it here. Trying to have the impl match the mental model of such a class (for reasoning / maintenance purposes mainly - much easier to verify things with clear semantics).

> Could I alternatively propose that we changed the name of the tensor to SimpleTensorDict? In that model this tensor is more like a wrapper over a dict, in which every field is seen uniformly. In this world, the current way of construction should be most natural.
>
> Thinking out side of the NJT context, perhaps a reason to favor SimpleTensorDict over CacheTensor is that
CacheTensor doesn't actually benefit much from being first and foremost a tensor, because it immediately unwraps upon any operators call. A container-like name of simple-tensor-dict makes it clearer that this is more of a container object than a tensor you are meant to operate on.

Okay I generally like where this is going- if you really want to act on a dictionary cache, use an object that fits that model. How would the torch_dispatch for `SimpleTensorDict` work - would it still unwrap for a specific key?

Reply:
That is a good point - It is true that the current behavior is not symmetric wrt all the field... maybe what I really want to happen is to ban all operations.

Reply:
> maybe what I really want to happen is to ban all operations.

This should be fine except for the purpose of compile, I need to support 'detach'. WDYT?

Reply:
> This should be fine except for the purpose of compile, I need to support 'detach'. WDYT?

Sure, so it's only purpose is to store tensors into a dict in a way that matches the "traceable wrapper subclass" interface? Does `detach()` detach all the tensors in the dict?

Reply:
> Sure, so it's only purpose is to store tensors into a dict in a way that matches the "traceable wrapper subclass" interface?

Exactly

>  Does detach() detach all the tensors in the dict?

I think we want to return a fresh NestedTensor object, but I'm not completely sure if I want to detach everything inside.

What tips me slightly toward not detaching overall is because generally for NT's size comparison logic, we'd prefer if the instance of the offsets, etc is the same. (I'm not sure it actually errors either when I do detach, but didn't  investigate further)

For other things it seems like it shouldn't matter:

- From the perspective of compile it seems okay not to detach because we already have the precedence of NJT not detaching offsets when the outer NJT is detached.

- From the perspective of autograd it shouldn't matter either here since detach is a no-op underneath autograd.

One might argue that one would expect a dict-container-like tensor replicate the outer commands into all the inner commands, which I don't disagree with, but at the same time, I don't really consider detach() truly public API either 
because I don't expect the user to be randomly calling into detach when no other op is supported.